samples/AspNetCore/SecretStoreConfigurationProviderSample/Program.cs

```diff
@@ -0,0 +1,93 @@
+﻿namespace SecretStoreConfigurationProviderSample
+{
+    using Microsoft.AspNetCore.Hosting;
+    using Microsoft.Extensions.Hosting;
+    using Dapr.Client;
+    using Dapr.Extensions.Configuration;
+    using Microsoft.Extensions.DependencyInjection;
+    using System.Net.Sockets;
+    using System;
+    using System.Net;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Secret Store Configuration Provider Sample.
+    /// </summary>
+    public class Program
+    {
+        /// <summary>
+        /// Main for Secret Store Configuration Provider Sample.
+        /// </summary>
+        /// <param name="args">Arguments.</param>
+        public static void Main(string[] args)
+        {
+            CreateHostBuilder(args).Build().Run();
+        }
+
+        /// <summary>
+        /// Creates WebHost Builder.
+        /// </summary>
+        /// <param name="args">Arguments.</param>
+        /// <returns>Returns IHostbuilder.</returns>
+        public static IHostBuilder CreateHostBuilder(string[] args)
+        {
+            WaitForDapper();
+
+            // Create Dapr Client
+            DaprClient client = new DaprClientBuilder().Build();
+            return Host.CreateDefaultBuilder(args)
+                .ConfigureServices((services) =>
+                {
+                    // Add the DaprClient to DI.
+                    services.AddSingleton<DaprClient>(client);
+                })
+                .ConfigureAppConfiguration((configBuilder) =>
+                {
+                    // Create descriptors for the secrets you want to rerieve from the Dapr Secret Store.
+                    var secretDescriptors = new DaprSecretDescriptor[]
+                    {
+                        new DaprSecretDescriptor(
+                            "super-secret",
+                            new Dictionary<string, string>(){ { "namespace", "dapr-tests" } }
+                        )
+                    };
+
+                    // Add the secret store Configuration Provider to the configuration builder.
+                    configBuilder.AddDaprSecretStore("kubernetes", secretDescriptors, client);
+                })
+                .ConfigureWebHostDefaults(webBuilder =>
+                {
+                    webBuilder.UseStartup<Startup>();
+                });
+        }
+
+        private static void WaitForDapper()
```

Comment:
@amanbha this is the solution I had to implement to trick Dapr to believe the app is ready and then to make the app wait for the gRPC port to be available.

Should we implement something in the SDK so Devs do not have to write this or another workaround en every app?

Reply:
Make app wait for gRPC port to be available seems fine to add to SDK.

Making Dapr believe that app is ready should not be required as the samples/Client works without it. Also when users write a real service, they would have a server in it.

Reply:
Opening the port is required cause Dapr pauses execution until the app is available and therefore gRPC port never gets open. Secrets need to be read before Startup kicks so the so the user server is not ready yet.

Reply:
I tried running the sample without a server in app.
`
E:\github\dapr\dotnet-sdk\samples\Client\DaprClient>dapr run dotnet run
Starting Dapr with id Healermarble-Crown. HTTP Port: 54240. gRPC Port: 54241
[93;1m== DAPR == time="2020-03-19T15:52:26.5285499-07:00" level=info msg="starting Dapr Runtime -- version 0.5.0 -- commit v0.3.0-rc.0-157-gb3bcd88-dirty" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5285499-07:00" level=info msg="log level set to: info" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5285499-07:00" level=info msg="metrics server started on :54242/" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.metrics type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5285499-07:00" level=info msg="standalone mode configured" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5285499-07:00" level=info msg="dapr id: Healermarble-Crown" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5285499-07:00" level=info msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5305485-07:00" level=info msg="found component messagebus (pubsub.redis)" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5305485-07:00" level=info msg="found component statestore (state.redis)" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == 2020-03-19 15:52:26.531548 I | redis: connecting to localhost:6379
[0m
[93;1m== DAPR == 2020-03-19 15:52:26.534581 I | redis: connected to localhost:6379 (localAddr: [::1]:54249, remAddr: [::1]:6379)
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5385818-07:00" level=info msg="Initialized service discovery to standalone" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5385818-07:00" level=error msg="failed to init input bindings: app channel not initialized" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5385818-07:00" level=info msg="actor runtime started. actor idle timeout: 1h0m0s. actor scan interval: 30s" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5395835-07:00" level=info msg="starting connection attempt to placement service at localhost:6050" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5395835-07:00" level=info msg="enabled metrics grpc middleware" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.grpc type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5395835-07:00" level=info msg="gRPC server is running on port 54241" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5395835-07:00" level=info msg="enabled cors http middleware" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.http type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5405818-07:00" level=info msg="enabled proxy http middleware" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.http type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5405818-07:00" level=info msg="enabled metrics http middleware" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.http type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5405818-07:00" level=info msg="http server is running on port 54240" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5405818-07:00" level=info msg="local service entry announced" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5405818-07:00" level=info msg="dapr initialized. Status: Running. Init Elapsed 12.0319ms" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime type=log ver=0.5.0
[0m
Updating metadata for app command: dotnet run
You're up and running! Both Dapr and your app logs will appear here.

[93;1m== DAPR == time="2020-03-19T15:52:26.5575835-07:00" level=info msg="established connection to placement service at localhost:6050" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5635825-07:00" level=info msg="placement order received: lock" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5635825-07:00" level=info msg="placement order received: update" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5635825-07:00" level=info msg="actors: placement tables updated" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[93;1m== DAPR == time="2020-03-19T15:52:26.5635825-07:00" level=info msg="placement order received: unlock" app_id=Healermarble-Crown instance=AMANDEV02 scope=dapr.runtime.actor type=log ver=0.5.0
[0m
[94;1m== APP == Published Event!
[0m
[94;1m== APP == Saved State!
[0m
[94;1m== APP == Got State: small  yellow
[0m
[94;1m== APP == Deleted State!
[0m


`

Reply:
For the given sample I run `dapr run --port 3500 --app-port 80 --app-id test dotnet run` which is needed for the web app (and the way it runs in k8s). If I don't open port 80 in the `WaitForDapper` method then Dapr keeps waiting here:

```shell
== DAPR == time="2020-03-20T12:00:00.8219418+01:00" level=info msg="application protocol: http. waiting on port 80" app_id=te
st instance=WFH scope=dapr.runtime type=log ver=0.5.0
```

Which prevents Dapr from listening gRPC calls  and therefore the app fails with the following exception:

 _Error starting gRPC call: No connecti
on could be made because the target machine actively refused it._

Reply:
@cmendible Got it. While waiting for Dapr in SDK makes sense. Listening for a port in Dapr SDK is not required as mostly users would have an endpoint in the service listening on it.